Plugin integration testing framework

elliothershberg opened the pull request:
# Overview

This PR introduces an integration testing framework for external JBrowse 2 plugins, as well as some additional developer conveniences.

## Testing

The approach taken was to create a dot directory (`.jbrowse/`) with a JB2 build. This local build can be used for testing the plugin being developed. It can also just be started on its own (`yarn browse`). When it is started, it uses `jbrowse_config.json` as its configuration, which makes the plugin available.

With a local build available, I configured [cypress](https://www.cypress.io/), which is a popular, well documented, and widely used integration testing framework. The cypress test runner can be run using `yarn cypress` which provides some really great features, or the tests can be run on their own (`yarn test` if you are in the middle of developing, or `yarn e2e` which automates the entire process for you).

## Developer conveniences

- Included a small plugin example in the index, with two examples of integration tests for it
- Added a setup script to automate the various manual configuration steps
- Added the yarn script `browse-install` which automates setting up the JB2 build for testing
- Created a GH action that runs the integration testing in CI when the repository is pushed to.

All of these changes are documented in the README.

elliothershberg commented:
Working out one kink with the setup step.

elliothershberg commented:
Alright, a bit cleaned up! 👍

rbuels commented:
this is cool! one question: the old template had `tsdx test` in it for running unit tests, what does this template use for unit tests?

elliothershberg commented:
@rbuels I did think about that, wasn't sure if we still needed that.

One think I've seen in various docs are commands like `test:cy` for cypress.

Could maybe add a command `"test:unit": tsdx test`

Or have `"test": tsdx test` and then `"test:cy" be the cypress integration testing

garrettjstevens reviewed (changes requested):
Could you add a check in `setup.js` to make sure that the `package.json` has a "name" field? That way we can be somewhat sure that they've already run `yarn init`.

garrettjstevens commented:
Also, maybe `setup.js` should generate the GitHub workflow file instead of it being in the template? That way it doesn't run and fail in this repository.

elliothershberg commented:
- Added back `tsdx test` for unit tests
- Added check for name in `package.json` during setup
- Now the github action isn't run in the template since it isn't in `.github/workflows` which Github Actions looks for. Gets moved there during setup to activate.
- Fixed issues with some files not having newlines

garrettjstevens commented on scripts/setup.js:
```diff
@@ -0,0 +1,104 @@
+}
+
+// 1. Change "name" in the "jbrowse-plugin" field to the name of your project (e.g. "MyProject")
+packageJSON['jbrowse-plugin'].name = projectName
```
Generally the `jbrowse-plugin` name is different from the package name. It should have any `@scope` and the "jbrowse-plugin-" stripped from it and be converted to PascalCase. This is also what should be used in the start and build scripts and the "name" of the plugin in any config file.

elliothershberg replied:
Fixed

garrettjstevens commented on scripts/setup.js:
```diff
@@ -0,0 +1,104 @@
+packageJSON.scripts.build = `tsdx build --format cjs,esm,umd --name JBrowsePlugin${projectName}`
+
+// 3. In the "module" field, replace jbrowse-plugin-my-project with the name of your project (leave off the @myscope if using a scoped package name) (you can double-check that the filename is correct after running the build step below and comparing the filename to the file in the dist/ folder)
+packageJSON.module = `dist/${projectName}.esm.js`
```
This should match what TSDX does, see here: https://github.com/formium/tsdx/blob/462af2d002987f985695b98400e0344b8f2754b7/src/utils.ts#L18-L21. It should also be used in the plugin "url" of any config file.

elliothershberg replied:
Fixed

garrettjstevens commented on scripts/setup.js:
```diff
@@ -0,0 +1,104 @@
+  -4,
+)}/workflows/Integration/badge.svg?branch=main)`
+const stream = fs.createWriteStream('README.md')
+README.forEach(line => stream.write(`${line}\r\n`))
```
Should use `\n' or `os.EOL` for line ending

elliothershberg replied:
Fixed

garrettjstevens commented on scripts/setup.js:
```diff
@@ -0,0 +1,104 @@
+  0,
+  -4,
+)}/workflows/Integration/badge.svg?branch=main)`
+const stream = fs.createWriteStream('README.md')
```
For consistency, it might be better to use `fs.writeFileSync('README.md', README.join('\n'), 'utf8')`

elliothershberg replied:
Fixed

garrettjstevens commented on package.json:
```diff
@@ -12,9 +12,15 @@
     "src"
   ],
   "scripts": {
+    "setup": "node scripts/setup.js && mkdir .jbrowse && yarn browse-install",
```
This assumes the user has yarn which they might not. `npm run` is safer.

Also, I'd like it if `yarn setup` didn't fail if run twice, it just skipped or re-did steps.

elliothershberg replied:
Fixed

elliothershberg replied:
For the question of yarn setup, I fixed it so that it can be run multiple times without breaking by moving making the directory into the setup script, and in the script check if it is already there before making it.

garrettjstevens commented on cypress/fixtures/hello_view.json:
```diff
@@ -0,0 +1,57 @@
+      "url": "http://localhost:9000/dist/jb2-plugin-testing.umd.development.js"
+    }
+  ],
+  "defaultSession": {
+    "id": "j8wx-zukm",
+    "name": "New session 1/12/2021, 12:32:05 PM",
+    "margin": 0,
+    "drawerWidth": 384,
+    "views": [
+      {
+        "type": "HelloView"
+      }
+    ],
+    "widgets": {},
+    "activeWidgets": {},
+    "connectionInstances": {},
+    "sessionTracks": [],
+    "sessionConnections": []
+  }
```
This can be simplified

```suggestion
  "defaultSession": {
    "name": "New session",
    "views": [
      {
        "type": "HelloView"
      }
    ],
  }
```

elliothershberg replied:
Fixed

garrettjstevens commented on src/index.tsx:
```diff
@@ -0,0 +1,56 @@
+    function ReactComponent() {
+      const [pushed, setPushed] = React.useState('')
+      return (
+        <div style={{ padding: 50 }}>
+          <h1>Hello plugin developers!</h1>
+          <button
+            onClick={() => {
+              setPushed('Woah! You pushed the button!')
+            }}
+          >
+            Push the button
+          </button>
+          <p>{pushed}</p>
+        </div>
```
Could you put the React components in a different file so this can remain a `.ts` and not a `.tsx` file/

elliothershberg replied:
Fixed

garrettjstevens commented on integration.yml:
```diff
@@ -0,0 +1,21 @@
+      - name: Get latest JBrowse
+        run: mkdir .jbrowse && yarn run jbrowse create --force .jbrowse
+      - name: Test plugin
+        run: yarn run e2e
```
I think it would be good if this ran `yarn test` as well, maybe `yarn test --passWithNoTests` to be safe.

elliothershberg replied:
Fixed. Tested this and make sure it actually passed on the GH action, see here: https://github.com/elliothershberg/jbrowse-plugin-template-test/runs/1772737341

garrettjstevens commented on README.md:
```diff
@@ -1,112 +1,137 @@
+plugin already included in the configuration. Your plugin must also be running
+(`yarn start` or `npm run start`).
+
+#### `yarn test` or `npm test`
```
I think this is `test:cy` now? also we should probably add a section on what `yarn test` does.

elliothershberg replied:
Fixed

elliothershberg commented on scripts/setup.js:
```diff
@@ -0,0 +1,104 @@
+
+// add status badge to README
+let README = readFile('README.md').split(/\r?\n/)
+README[0] = `# ${projectName} ![Integration](${packageJSON.repository.slice(
```
Made a more robust function for retrieving url

elliothershberg commented on scripts/setup.js:
```diff
@@ -0,0 +1,104 @@
+
+function writeJSON(data, path) {
+  try {
+    fs.writeFileSync(path, JSON.stringify(data, null, 2))
```
Fixed 👍

elliothershberg commented on cypress/fixtures/hello_view.json:
```diff
@@ -0,0 +1,57 @@
+{
```
Fixed

elliothershberg commented on scripts/setup.js:
```diff
@@ -0,0 +1,104 @@
+const fs = require('fs')
+const path = require('path')
+
+/* 
```
Linted this file, and ultimately removed those comments. This should be fixed. How did you find those? Was that from running lint locally?

garrettjstevens replied:
My editor is set to strip trailing whitespace on save, so I noticed it when I opened the file in my editor and saw them get removed automatically.

garrettjstevens commented on scripts/setup.js:
```diff
@@ -0,0 +1,178 @@
+  const repoUrl = parseRepoUrl(packageJSON.repository)
+  if (repoUrl !== undefined) {
+    let README = readFile('README.md').split(/\r?\n/)
+    README[0] = `# ${projectName} ![Integration](${repoUrl}/workflows/Integration/badge.svg?branch=main)`
```
I think the badge should go on the top line, then a blank line, then the top-level header. Also, I think the top-level header should get re-written even if the there is no `repoUrl`.

garrettjstevens commented on scripts/setup.js:
```diff
@@ -0,0 +1,178 @@
+    }
+  }
+
+  if (typeof url === 'string' && url.includes('github.com')) {
```
The includes check is done on the next line now.

```suggestion
  if (typeof url === 'string')) {
```

garrettjstevens commented on README.md:
```diff
@@ -1,112 +1,141 @@
+
+Runs any unit tests defined during plugin development.
+
+#### `yarn test:cy` or `npm test:cy`
```
NPM command needs "run"

```suggestion
#### `yarn test:cy` or `npm run test:cy`
```

garrettjstevens commented on scripts/setup.js:
```diff
@@ -0,0 +1,178 @@
+  }
+}
+
+/* 
```
Trailing whitespace here

```suggestion
/*
```

garrettjstevens commented on README.md:
```diff
@@ -1,112 +1,141 @@
+
+Now run `yarn` (or `rm yarn.lock && npm install` to use npm instead of yarn) to install the necessary dependencies.
+
+After this, run `yarn setup` (or `npm run setup`). 
```
Trailing whitespace here

```suggestion
After this, run `yarn setup` (or `npm run setup`).
```

garrettjstevens commented on README.md:
```diff
@@ -1,112 +1,141 @@
+
+#### `yarn e2e` or `npm run e2e`
+
+Starts up the JBrowse 2 build as well as your plugin, and runs the [cypress](https://www.cypress.io/) 
```
Trailing whitespace here

```suggestion
Starts up the JBrowse 2 build as well as your plugin, and runs the [cypress](https://www.cypress.io/)
```

garrettjstevens commented on README.md:
```diff
@@ -1,112 +1,141 @@
+#### `yarn cypress` or `npm run cypress`
+
+Launches the [cypress](https://www.cypress.io/) test runner, which can be very
+useful for writing integration tests for your plugin. Both the plugin and `browse` 
```
Trailing whitespace here

```suggestion
useful for writing integration tests for your plugin. Both the plugin and `browse`
```
